sklearn/cluster/_optics.py

```diff
@@ -609,12 +615,17 @@ def _set_reach_dist(
             # the same logic as neighbors, p is ignored if explicitly set
             # in the dict params
             _params["p"] = p
-        dists = pairwise_distances(
-            P, np.take(X, unproc, axis=0), metric=metric, n_jobs=None, **_params
-        ).ravel()
+        dists = pairwise_distances(P, X[unproc], metric, n_jobs=None, **_params).ravel()
 
-    rdists = np.maximum(dists, core_distances_[point_index])
-    np.around(rdists, decimals=np.finfo(rdists.dtype).precision, out=rdists)
+    if issparse(dists):
+        rdists = dists.maximum(core_distances_[point_index])
```

Comment:
Scipy warns here that max will result in a dense matrix. Since we convert to dense anyway it would make sense to convert before this. I'd convert at line 611:
```py
    if metric == "precomputed":
        dists = X[point_index, unproc]
        if issparse(dists):
            dists = dists.toarray().ravel()
```

Then, this branch is no longer useful